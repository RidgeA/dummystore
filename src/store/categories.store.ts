import {createAsyncThunk, createSlice} from "@reduxjs/toolkit";
import {RootState} from "./store";

export interface Category {
    id: string;
    displayName: string;
}

interface CategoriesState {
    list: Category[];
    status: "idle" | "loading" | "succeeded" | "error"
    error?: string
}

const allCategories: Category = {
    id: "",
    displayName: "All categories"
}

const initialState: CategoriesState = {
    list: [allCategories],
    status: "idle",
}

export const categoriesSlice = createSlice({
    name: 'categories',
    initialState,
    reducers: {},
    extraReducers(builder) {
        builder
            .addCase(loadCategories.pending, (state,) => {
                state.status = "loading";
            })
            .addCase(loadCategories.fulfilled, (state, action) => {
                state.status = "succeeded"
                state.list = [allCategories, ...action.payload.map(transformCategory)]
            })
            .addCase(loadCategories.rejected, (state, action) => {
                state.status = "error"
                state.error = action.error.message
            })
    }
})


export const selectAllCategory = (state: RootState) => state.categories.list

export const loadCategories = createAsyncThunk("categories/load", async () => {
    const res = await fetch("https://dummyjson.com/products/categories")
    return res.json()
})

function transformCategory(category: string): Category {
    return {
        id: category,
        displayName: category.split("-").map(s => s.charAt(0).toUpperCase() + s.slice(1)).join(" ")
    }
}