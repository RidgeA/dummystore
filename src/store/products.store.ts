import {createAsyncThunk, createSlice, Draft} from "@reduxjs/toolkit";
import {RootState} from "./store";

export interface Product {
    id: number
    title: string
    description: string
    price: number
    category: string
    thumbnail: string
    images: string[]
    rating: number
}

interface ProductsState {
    list: Product[]
    page: { current: number; total: number };
    limit: number
    status: "idle" | "loading" | "succeeded" | "error"
    error?: string
    product?: Product,
    searchTerm: string
}

export interface ProductsParameters {
    categoryId?: string
    page?: number
    search?: string
}

const initialState: ProductsState = {
    list: [],
    page: {current: 0, total: 0},
    limit: 10,
    status: "idle",
    searchTerm: ""
}

export const productsSlice = createSlice({
    name: 'products',
    initialState,
    reducers: {},
    extraReducers(builder) {
        builder
            .addCase(loadProducts.pending, loadingReducer)
            .addCase(loadProducts.rejected, errorReducer)
            .addCase(loadProducts.fulfilled, successReducer)

            .addCase(loadProductsWithParameters.pending, loadingReducer)
            .addCase(loadProductsWithParameters.rejected, errorReducer)
            .addCase(loadProductsWithParameters.fulfilled, successReducer)

            .addCase(loadProductById.pending, loadingReducer)
            .addCase(loadProductById.rejected, errorReducer)
            .addCase(loadProductById.fulfilled, (state, action) => {
                state.status = "succeeded"
                state.product = action.payload
            })
    }
})

const loadingReducer = (state: Draft<ProductsState>) => {
    state.error = undefined;
    state.status = "loading"
}

const errorReducer = (state: Draft<ProductsState>, action: any) => {
    state.status = "error"
    state.error = action.error.message
}

const successReducer = (state: Draft<ProductsState>, action: any) => {
    state.status = "succeeded"
    const {products: list, total, skip, limit} = action.payload
    state.list = list
    state.page.total = total / limit
    state.page.current = (skip / limit) + 1
}

export const loadProducts = createAsyncThunk<any, void, { state: RootState }>("products/load", async (params, api) => {
    const {products: {limit}} = api.getState()
    const res = await fetch(`https://dummyjson.com/products?limit=${limit}`)
    return res.json()
})

export const loadProductsWithParameters = createAsyncThunk<any, ProductsParameters, { state: RootState }>("products/load/byCategory", async (params, api) => {

    const {products: {limit}} = api.getState()

    const skip = ((params.page || 1) - 1) * limit

    const qps = [
        `limit=${limit}`,
        `skip=${skip}`,
    ].join("&")

    let res;
    switch (true) {
        case !!params.categoryId: {
            res = await fetch(`https://dummyjson.com/products/category/${params.categoryId}?${qps}`)
            break;
        }
        case !!params.search: {
            res = await fetch(`https://dummyjson.com/products/search?q=${params.search}&${qps}`)
            break;
        }
        default:
            res = await fetch(`https://dummyjson.com/products?${qps}`)
    }

    return res.json()
})

export const loadProductById = createAsyncThunk<any, number, { state: RootState }>("products/load/byId", async (productId: number) => {
    const res = await fetch(`https://dummyjson.com/products/${productId}`)
    return res.json()
})