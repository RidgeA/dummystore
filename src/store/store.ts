import {combineReducers, configureStore, PreloadedState} from "@reduxjs/toolkit";
import {categoriesSlice} from "./categories.store";
import {productsSlice} from "./products.store";

const rootReducer = combineReducers({
    categories: categoriesSlice.reducer,
    products: productsSlice.reducer,
})

export const setupStore = (preloadedState?: PreloadedState<RootState>) => {
    return configureStore({
        reducer: rootReducer,
        preloadedState
    })
}

export type RootState = ReturnType<typeof rootReducer>
export type AppStore = ReturnType<typeof setupStore>
export type AppDispatch = AppStore['dispatch']
