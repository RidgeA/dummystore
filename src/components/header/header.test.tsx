import * as TestRenderer from 'react-test-renderer';
import Header from './header';
import React from "react";
import {testRouter} from "../../utils/test/router";
import {setupStore} from "../../store/store";
import {testStore} from "../../utils/test/store";

describe('Header', () => {
    it('renders', () => {
        const tree = TestRenderer
            .create(
                testStore(testRouter(<Header title="My Title"></Header>), setupStore())
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
});