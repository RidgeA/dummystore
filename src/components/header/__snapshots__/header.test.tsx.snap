// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header renders 1`] = `
<nav
  className="navbar navbar-expand-lg navbar-light bg-light"
>
  <div
    className="container-fluid"
  >
    <span
      className="navbar-brand"
    >
      My Title
    </span>
    <div
      aria-expanded={null}
      className="navbar-collapse collapse"
    >
      <div
        className="ma-auto navbar-nav"
        onKeyDown={[Function]}
      >
        <span
          className="nav-link"
          data-rr-ui-event-key={null}
          disabled={false}
          onClick={[Function]}
        >
          <a
            href="/"
            onClick={[Function]}
          >
            Catalog
          </a>
        </span>
        <span
          className="nav-link"
          data-rr-ui-event-key={null}
          disabled={false}
          onClick={[Function]}
        >
          <a
            href="/order"
            onClick={[Function]}
          >
            Order & delivery
          </a>
        </span>
        <span
          className="nav-link"
          data-rr-ui-event-key={null}
          disabled={false}
          onClick={[Function]}
        >
          <a
            href="/about"
            onClick={[Function]}
          >
            About Us
          </a>
        </span>
      </div>
    </div>
    <form
      className=""
      onSubmit={[Function]}
    >
      <div
        className="input-group"
      >
        <input
          className="search-input form-control"
          onChange={[Function]}
          placeholder="Search"
          type="text"
          value=""
        />
        <button
          className="btn btn-primary search-button"
          type="submit"
        >
          üîç
        </button>
      </div>
    </form>
  </div>
</nav>
`;
